### PR TITLE
Custom pre-commit hooks for our monorepo setup

### PR DESCRIPTION
**Pull request by szg251**
- Configuring nix for rust prelude
- Remove rust test from checks
- Naive implementation of JSON serialisers/deserialisers in Rust
- Removing cargo workspaces
- Move rust specific gitignores to package directory
- Implement JSON roundtrip tests
- Experimenting with flakes modules
- Remove crates.nix
- Delete proptest-regressions file
- Add documentation to exposed functions
- Remove 'import' statement for flake module imports
- Add rustfmt and clippy

### CODE REVIEW

File: .envrc [hunk 1]
```diff
@@ -1 +1,2 @@
+nix_direnv_watch_file ./pre-commit.nix
```
Comment (bladyjoker) [inline .envrc:1]: Was is das? Just curious :)

Comment (szg251) [reply .envrc:1]: I think I mentioned this in our session, we can set direnv to reload automatically, when some files are changed (this takes multiple arguments)

File: pre-commit.nix [hunk 2]
```diff
@@ -1,14 +1,15 @@
-{ inputs, lib, ... }: {
+{ inputs, ... }: {
   imports = [
-    inputs.pre-commit-hooks.flakeModule # Adds perSystem.pre-commit options
+    ./extras/pre-commit-hooks-extra.nix
```
[the remark this answers is not in the document]
Comment (szg251) [reply pre-commit.nix:3]: I think we covered this as well..

**Review by bladyjoker — APPROVED** [review]
I think you got this :+1: 

I'm happy about `deadnix` :D

File: extras/pre-commit-hooks-extra.nix [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  ];
+  perSystem = { pkgs, ... }:
+    let
+      rustfmt = pkgs.rust-bin.stable.latest.rustfmt;
+    in
+    {
+      pre-commit.settings.hooks = {
+        rustfmt-monorepo =
+          {
+            name = "rustfmt";
```
Comment (szg251) [inline extras/pre-commit-hooks-extra.nix:13]: Ended up with a ridiculously simple configuration, now that clippy is out of the picture

**Pull request merged**